New comparison data

ajparsons opened the pull request:
Work in progress pull request for similar council comparisons. 

Will need `script/manage migrate` and `script/manage add_related_councils` to update a current system. 

Otherwise is run as part of regular setup. 

This has a very basic display on the council page, and support for searching on a limited range of councils, but nothing in the search page reflecting that is what's happening.

struan commented on caps/management/commands/add_related_councils.py:
```diff
@@ -1,41 +1,49 @@
+            print("downloading data")
+            download_data()
+        print('adding related authorities')
+        add_related_councils()
```
This will never update the data after the first time (without `get_all`) but will completely re-import everything into the database each time. I think `add_related_councils` should also be inside the if otherwise we're doing a lot of unnecessary database things on each update.

ajparsons replied:
Yep, indented so everything is within the all (although this now does nothing without the 'all' , but maybe it will in future?)

struan reviewed (changes requested):
I've not gone through the import code yet but I think the data fetching might need a different approach.

ajparsons commented on caps/management/commands/add_related_councils.py:
```diff
@@ -1,41 +1,49 @@
+    dst = Path("data", "comparisons")
+    dst.mkdir(exist_ok=True)
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
```
How about if it was fetching the latest release zip from github rather than cloning it?

struan replied:
oh, yes, that would be good. there's a python zip library too which means it could do all that in python

ajparsons replied:
This has been replaced by a zip file, so should now work ok without git.

struan commented on caps/forms.py:
```diff
@@ -26,9 +31,22 @@ def search(self):
+                             position__lte=cut_off)
+                     .values_list("council_b__name", flat=True))
+            
+            sqs = sqs.filter(council_name__in=names)
```
Using filter means we include the council name in the highlights which I don't think is what we want. This seems to do the same:
```
query_names = [ 'council_name:"{}"'.format(name) for name in names ]
            sqs = sqs.narrow(" OR ".join(query_names))
```

ajparsons replied:
That makes sense. Have done it with an f-string, but that works fine.

struan reviewed (changes requested):
Lots of small things but nothing substantial and it all works as far as I can tell, mostly about clarity.

ajparsons commented on caps/models.py:
```diff
@@ -110,6 +110,58 @@ def __str__(self):
+
+        processed_councils = []
+        for d in self.distances.all():
+            nc = deepcopy(council_lookup[d.council_b_id])
```
This is making it create a new object that avoids any of its properties being references that might be shared between multiple instances. 

The situation I want is that the same council may appear in multiple list of comparisons depending on the comparison type.

This *might* not be necessary, but I think there's a risk that `processed_councils` may end up containing multiple councils that are just references to the same object.

What I'm trying to avoid is a situation where the next line is assigning a distance to the council object, but the same council may come up in a separate comparison, be referenced again, and have a new distance assigned to it.

ajparsons replied:
Added

ajparsons commented on caps/models.py:
```diff
@@ -110,6 +110,58 @@ def __str__(self):
+                              earliest_promise=Min('promise__target_year'),
+                              declared_emergency=Min('emergencydeclaration__date_declared')))
+
+        council_lookup = {x.id: x for x in councils}
```
👍

ajparsons commented on caps/models.py:
```diff
@@ -110,6 +110,58 @@ def __str__(self):
+        types = {x.id: x for x in types}
+
+        key = lambda x: x.distance.type_id
+        processed_councils.sort(key=key)
```
Have done that and moved the lambda into the groupby function.

ajparsons commented on caps/models.py:
```diff
@@ -110,6 +110,58 @@ def __str__(self):
+        grouped_labels = groupby(labels, lambda x: x.label.type_id)
+
+        # change list of labels to tiered lookup (type_id, council_id)
+        labels = {k: {y.council_id: y for y in v} for k, v in grouped_labels}
```
Yep that's nicer, changed.

struan reviewed (approved):
👍

zarino commented:
I’ve tidied up the frontend for this, with helpful input from @ajparsons.

<img width="888" alt="Screenshot 2021-12-22 at 18 26 41" src="https://user-images.githubusercontent.com/739624/147138525-7cb4ab53-4795-42e4-9074-5d61c1a6231b.png">

@ajparsons is there any chance you could either update or temporarily remove the "(add link later)" bit at the end of `data/comparisons/emissions/description.md`? I’m guessing you put that there, hoping your blog post would be published before / at the same time that this was deployed?

@struan I think, aside from that "(add link later)" fix, everything here is good to go from my perspective.

zarino reviewed (approved):
Aside from the "(add link later)" that gets downloaded to `data/comparisons/emissions/description.md` (and then displayed on `/about/data`) – I’m good with everything else here :+1:

ajparsons commented:
@Zarino, checking out that branch I don't get the shortened names, where did you make that change?

![image](https://user-images.githubusercontent.com/8157058/147224563-d6a012ff-eb68-4efe-934d-eaa069ad2cab.png)

(Also, currently failing the linting test - there's some new instructions on the readme about how to enforce that)

zarino commented:
> @zarino, checking out that branch I don't get the shortened names, where did you make that change?

I shortened the names in `caps/models.py`, but the names are only read _once_ from that model, during import, and then saved in the database. To reveal the new, shorter names, in the UI, you’ll need to reimport all the data:

```
docker compose exec app ./manage.py add_related_councils --all
```

ajparsons commented:
Doh. Yep, that's fine. Adjusted the source dataset and incremented the version used by CAPE to change the description. Also fixed that bug where scores were out of order.
